### PR TITLE
Add new grpc lb plugin 'xds'

### PR DESCRIPTION
**Pull request by vishalpowar**
This plugin currently implements the grpclb protocol.

NOTE: All the files under xds/ directory have been cloned from grpclb/ and
will be updated in subsequently to support xds API.

### CODE REVIEW

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @vishalpowar isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
CLA GitHub bot_

**Comment by vishalpowar** [comment]
fixed CLA

**Comment by vishalpowar** [comment]
Checking CLA update

**Comment by vishalpowar** [comment]
Checking CLA update

File: BUILD [hunk 1]
```diff
@@ -1265,6 +1265,73 @@ grpc_cc_library(
     ],
 )
 
+grpc_cc_library(
```
Comment (markdroth) [inline BUILD]: Please also update the `build.yaml` file and then run `./tools/buildgen/generate_projects.sh`.

File: src/core/ext/filters/client_channel/lb_policy/xds/xds.cc [hunk 2]
```diff
@@ -0,0 +1,1902 @@
...
+
+// Destroy function used when embedding client stats in call context.
+void DestroyClientStats(void* arg) {
+  static_cast<XdsLbClientStats*>(arg)->Unref();
+}
+
+void XdsLb::PendingPickSetMetadataAndContext(PendingPick* pp) {
+  /* if connected_subchannel is nullptr, no pick has been made by the RR
+   * policy (e.g., all addresses failed to connect). There won't be any
+
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/lb_policy/xds/xds.cc]: s/grpclb/xds/

File: src/core/ext/filters/client_channel/lb_policy/xds/xds.cc [hunk 3]
```diff
@@ -0,0 +1,1902 @@
...
+  // Update fallback address list.
+  if (fallback_backend_addresses_ != nullptr) {
+    grpc_lb_addresses_destroy(fallback_backend_addresses_);
+  }
+  fallback_backend_addresses_ = ExtractBackendAddresses(addresses);
+  // Make sure that GRPC_ARG_LB_POLICY_NAME is set in channel args,
+  // since we use this to trigger the client_load_reporting filter.
+  static const char* args_to_remove[] = {GRPC_ARG_LB_POLICY_NAME};
+  grpc_arg new_arg = grpc_channel_arg_string_create(
+      (char*)GRPC_ARG_LB_POLICY_NAME, (char*)"grpclb");
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/lb_policy/xds/xds.cc]: s/grpclb/xds/

File: src/core/ext/filters/client_channel/lb_policy/xds/xds.cc [hunk 4]
```diff
@@ -0,0 +1,1902 @@
...
+  GRPC_CLOSURE_INIT(&lb_channel_on_connectivity_changed_,
+                    &XdsLb::OnBalancerChannelConnectivityChangedLocked, this,
+                    grpc_combiner_scheduler(args.combiner));
+  GRPC_CLOSURE_INIT(&on_rr_connectivity_changed_,
+                    &XdsLb::OnRoundRobinConnectivityChangedLocked, this,
+                    grpc_combiner_scheduler(args.combiner));
+  GRPC_CLOSURE_INIT(&on_rr_request_reresolution_,
+                    &XdsLb::OnRoundRobinRequestReresolutionLocked, this,
+                    grpc_combiner_scheduler(args.combiner));
+  grpc_connectivity_state_init(&state_tracker_, GRPC_CHANNEL_IDLE, "grpclb");
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/lb_policy/xds/xds.cc]: s/grpclb/xds/

File: src/core/ext/filters/client_channel/lb_policy/xds/xds.h [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+#ifndef GRPC_CORE_EXT_FILTERS_CLIENT_CHANNEL_LB_POLICY_XDS_XDS_H
+#define GRPC_CORE_EXT_FILTERS_CLIENT_CHANNEL_LB_POLICY_XDS_XDS_H
+
+#include <grpc/support/port_platform.h>
+
+/** Channel arg indicating if a target corresponding to the address is grpclb
+ * loadbalancer. The type of this arg is an integer and the value is treated as
+ * a bool. */
+#define GRPC_ARG_ADDRESS_IS_XDS_LOAD_BALANCER \
+  "grpc.address_is_grpclb_load_balancer"
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/lb_policy/xds/xds.h]: s/grpclb/xds/

File: src/core/ext/filters/client_channel/lb_policy/xds/xds.h [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+/** Channel arg indicating if a target corresponding to the address is a backend
+ * received from a balancer. The type of this arg is an integer and the value is
+ * treated as a bool. */
+#define GRPC_ARG_ADDRESS_IS_BACKEND_FROM_XDS_LOAD_BALANCER \
+  "grpc.address_is_backend_from_grpclb_load_balancer"
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/lb_policy/xds/xds.h]: s/grpclb/xds/

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

File: src/core/ext/filters/client_channel/lb_policy/xds/xds.h (hunk 6 above)
Comment (vishalpowar) [reply src/core/ext/filters/client_channel/lb_policy/xds/xds.h]: done

File: src/core/ext/filters/client_channel/lb_policy/xds/xds.h (hunk 5 above)
Comment (vishalpowar) [reply src/core/ext/filters/client_channel/lb_policy/xds/xds.h]: done

File: src/core/ext/filters/client_channel/lb_policy/xds/xds.cc (hunk 4 above)
Comment (vishalpowar) [reply src/core/ext/filters/client_channel/lb_policy/xds/xds.cc]: done

File: src/core/ext/filters/client_channel/lb_policy/xds/xds.cc (hunk 3 above)
Comment (vishalpowar) [reply src/core/ext/filters/client_channel/lb_policy/xds/xds.cc]: done

File: src/core/ext/filters/client_channel/lb_policy/xds/xds.cc (hunk 2 above)
Comment (vishalpowar) [reply src/core/ext/filters/client_channel/lb_policy/xds/xds.cc]: done

File: BUILD (hunk 1 above)
Comment (vishalpowar) [reply BUILD]: done

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,965,922      Total (=)      1,965,922

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,909,004      Total (>)     10,908,997

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: BUILD (hunk 1 above)
Comment (markdroth) [reply BUILD]: Still need to run `./tools/buildgen/generate_projects.sh` to regenerate the other build files.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,965,922      Total (=)      1,965,922

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,909,000      Total (<)     10,909,001

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,965,242      Total (=)      1,965,242

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,909,024      Total (<)     10,909,025

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by yashykt** [comment]
Known issues : #16846

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] No significant performance differences
```

**Pull request merged**